File: rust/src/engine/turbo/thrift.rs
//! Just enough of Thrift's compact protocol to read a Parquet footer.
//!
//! Parquet's metadata is a Thrift structure, and a Thrift library would bring a
//! code generator and a dependency for what is, at this scale, a few hundred
//! lines: field headers, zigzag varints, and the rule for skipping a field this
//! reader does not care about. Skipping correctly is the part that matters —
//! writers add fields, and a reader that cannot step over an unknown one breaks
//! on files it has never seen.

use crate::error::{Error, Result};

/// Thrift compact type codes, as they appear in the low nibble of a field header.
pub(super) mod kind {
    pub(in super::super) const STOP: u8 = 0;
    pub(in super::super) const TRUE: u8 = 1;
    pub(in super::super) const FALSE: u8 = 2;
    pub(in super::super) const BYTE: u8 = 3;
    pub(in super::super) const I16: u8 = 4;
    pub(in super::super) const I32: u8 = 5;
    pub(in super::super) const I64: u8 = 6;
    pub(in super::super) const DOUBLE: u8 = 7;
    pub(in super::super) const BINARY: u8 = 8;
    pub(in super::super) const LIST: u8 = 9;
    pub(in super::super) const SET: u8 = 10;
    pub(in super::super) const MAP: u8 = 11;
    pub(in super::super) const STRUCT: u8 = 12;
}

/// A cursor over the bytes of a compact-protocol message.
pub(super) struct Reader<'a> {
    data: &'a [u8],
    at: usize,
    /// Field ids are deltas from the previous one within a struct, so the reader
    /// carries the last id seen and stacks it across nested structs.
    last_id: i16,
    stack: Vec<i16>,
}

/// One field header: which field, and what it holds.
pub(super) struct FieldHeader {
    pub(super) id: i16,
    pub(super) kind: u8,
}

fn short(what: &str) -> Error {
    Error::new(format!("the Parquet metadata ends in the middle of {what}"))
}

impl<'a> Reader<'a> {
    pub(super) fn new(data: &'a [u8]) -> Self {
        Reader {
            data,
            at: 0,
            last_id: 0,
            stack: Vec::new(),
        }
    }

    fn byte(&mut self) -> Result<u8> {
        let b = *self.data.get(self.at).ok_or_else(|| short("a value"))?;
        self.at += 1;
        Ok(b)
    }

    fn varint(&mut self) -> Result<u64> {
        let mut out = 0u64;
        let mut shift = 0u32;
        loop {
            let b = self.byte()?;
            out |= ((b & 0x7f) as u64) << shift;
            if b & 0x80 == 0 {
                return Ok(out);
            }
            shift += 7;
            if shift > 63 {
                return Err(Error::new(
                    "a Parquet metadata varint is longer than 64 bits",
                ));
            }
        }
    }

    fn zigzag(&mut self) -> Result<i64> {
        let n = self.varint()?;
        Ok(((n >> 1) as i64) ^ -((n & 1) as i64))
    }

    /// Enters a struct. Every `struct_begin` needs its `struct_end`, because
    /// field ids are relative to the enclosing struct's last id.
    pub(super) fn struct_begin(&mut self) {
        self.stack.push(self.last_id);
        self.last_id = 0;
    }

    pub(super) fn struct_end(&mut self) {
        self.last_id = self.stack.pop().unwrap_or(0);
    }

    /// The next field of the current struct, or `None` at its stop byte.
    pub(super) fn field(&mut self) -> Result<Option<FieldHeader>> {
        let header = self.byte()?;
        if header == kind::STOP {
            return Ok(None);
        }
        let kind = header & 0x0f;
        let delta = (header >> 4) as i16;
        let id = if delta == 0 {
            self.zigzag()? as i16
        } else {
            self.last_id + delta
        };
        self.last_id = id;
        Ok(Some(FieldHeader { id, kind }))
    }

    pub(super) fn i32(&mut self) -> Result<i32> {
        Ok(self.zigzag()? as i32)
    }

    pub(super) fn i64(&mut self) -> Result<i64> {
        self.zigzag()
    }

    pub(super) fn binary(&mut self) -> Result<&'a [u8]> {
        let len = self.varint()? as usize;
        let from = self.at;
        let to = from
            .checked_add(len)
            .filter(|&to| to <= self.data.len())
            .ok_or_else(|| short("a string"))?;
        self.at = to;
        Ok(&self.data[from..to])
    }

    pub(super) fn string(&mut self) -> Result<String> {
        Ok(String::from_utf8_lossy(self.binary()?).into_owned())
    }

    /// A list header: how many elements, and of what.
    pub(super) fn list_begin(&mut self) -> Result<(usize, u8)> {
        let header = self.byte()?;
        let kind = header & 0x0f;
        let size = (header >> 4) as usize;
        let size = if size == 15 {
            self.varint()? as usize
        } else {
            size
        };
        Ok((size, kind))
    }

    /// Steps over a value of `kind` without interpreting it. This is what lets a
    /// file written by a newer library, carrying fields this reader has never
    /// heard of, still be read.
    pub(super) fn skip(&mut self, kind: u8) -> Result<()> {
        match kind {
            kind::TRUE | kind::FALSE => Ok(()),
            kind::BYTE => self.byte().map(|_| ()),
            kind::I16 | kind::I32 | kind::I64 => self.zigzag().map(|_| ()),
            kind::DOUBLE => {
                self.at = self
                    .at
                    .checked_add(8)
                    .filter(|&to| to <= self.data.len())
                    .ok_or_else(|| short("a double"))?;
                Ok(())
            }
            kind::BINARY => self.binary().map(|_| ()),
            kind::LIST | kind::SET => {
                let (n, element) = self.list_begin()?;
                for _ in 0..n {
                    self.skip(element)?;
                }
                Ok(())
            }
            kind::MAP => {
                let n = self.varint()? as usize;
                if n == 0 {
                    return Ok(());
                }
                let kinds = self.byte()?;
                let (key, value) = (kinds >> 4, kinds & 0x0f);
                for _ in 0..n {
                    self.skip(key)?;
                    self.skip(value)?;
                }
                Ok(())
            }
            kind::STRUCT => {
                self.struct_begin();
                while let Some(f) = self.field()? {
                    self.skip(f.kind)?;
                }
                self.struct_end();
                Ok(())
            }
            other => Err(Error::new(format!(
                "the Parquet metadata holds a Thrift type this reader does not know: {other}"
            ))),
        }
    }

    /// Reads a whole struct, handing each field to `on_field`, which returns
    /// `false` for a field it did not consume so this reader can skip it.
    pub(super) fn read_struct<F>(&mut self, mut on_field: F) -> Result<()>
    where
        F: FnMut(&mut Self, &FieldHeader) -> Result<bool>,
    {
        self.struct_begin();
        while let Some(f) = self.field()? {
            if !on_field(self, &f)? {
                self.skip(f.kind)?;
            }
        }
        self.struct_end();
        Ok(())
    }

    /// Reads a list of structs, handing each to `each`.
    pub(super) fn read_list<F>(&mut self, mut each: F) -> Result<()>
    where
        F: FnMut(&mut Self) -> Result<()>,
    {
        let (n, element) = self.list_begin()?;
        for _ in 0..n {
            if element == kind::STRUCT {
                each(self)?;
            } else {
                self.skip(element)?;
            }
        }
        Ok(())
    }

    /// How far in the cursor has got, which a page header reader needs to know
    /// where the page's own bytes begin.
    pub(super) fn position(&self) -> usize {
        self.at
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// `{1: 7, 2: "hi", 3: [4, 5]}` in compact form, then a stop byte.
    fn message() -> Vec<u8> {
        vec![
            0x15, 0x0e, // field 1 (delta 1), i32, zigzag 14 -> 7
            0x18, 0x02, b'h', b'i', // field 2, binary, length 2
            0x19, 0x25, 0x08, 0x0a, // field 3, list of two i32: 4 and 5
            0x00,
        ]
    }

    #[test]
    fn fields_are_read_by_delta_and_values_unzigzagged() {
        let bytes = message();
        let mut r = Reader::new(&bytes);
        let mut seen = Vec::new();
        r.read_struct(|r, f| {
            match f.id {
                1 => seen.push(format!("i32 {}", r.i32()?)),
                2 => seen.push(format!("str {}", r.string()?)),
                3 => {
                    let (n, _) = r.list_begin()?;
                    let mut all = Vec::new();
                    for _ in 0..n {
                        all.push(r.i32()?);
                    }
                    seen.push(format!("list {all:?}"));
                }
                _ => return Ok(false),
            }
            Ok(true)
        })
        .expect("a well-formed message");
        assert_eq!(seen, ["i32 7", "str hi", "list [4, 5]"]);
    }

    #[test]
    fn a_field_nobody_wants_is_stepped_over() {
        let bytes = message();
        let mut r = Reader::new(&bytes);
        let mut last = None;
        r.read_struct(|r, f| {
            if f.id == 3 {
                let (n, _) = r.list_begin()?;
                let mut all = Vec::new();
                for _ in 0..n {
                    all.push(r.i32()?);
                }
                last = Some(all);
                return Ok(true);
            }
            Ok(false) // skipped, whatever it was
        })
        .expect("skipping to work");
        assert_eq!(last, Some(vec![4, 5]));
    }
}

File: zig/src/thrift.zig
//! Just enough of Thrift's compact protocol to read a Parquet footer.
//!
//! Parquet's metadata is a Thrift structure, and a Thrift library would bring a
//! code generator and a dependency for what is, at this scale, a few hundred
//! lines: field headers, zigzag varints, and the rule for skipping a field this
//! reader does not care about. Skipping correctly is the part that matters --
//! writers add fields, and a reader that cannot step over an unknown one breaks
//! on files it has never seen.

const std = @import("std");

pub const Error = error{ TruncatedMetadata, UnknownThriftType };

/// Thrift compact type codes, as they appear in the low nibble of a field header.
pub const STOP: u8 = 0;
pub const TRUE: u8 = 1;
pub const FALSE: u8 = 2;
pub const BYTE: u8 = 3;
pub const I16: u8 = 4;
pub const I32: u8 = 5;
pub const I64: u8 = 6;
pub const DOUBLE: u8 = 7;
pub const BINARY: u8 = 8;
pub const LIST: u8 = 9;
pub const SET: u8 = 10;
pub const MAP: u8 = 11;
pub const STRUCT: u8 = 12;

/// One field header: which field, and what it holds.
pub const FieldHeader = struct { id: i16, kind: u8 };

/// A cursor over the bytes of a compact-protocol message.
pub const Reader = struct {
    data: []const u8,
    at: usize = 0,
    /// Field ids are deltas from the previous one within a struct, so the reader
    /// carries the last id seen and stacks it across nested structs.
    last_id: i16 = 0,
    stack: [16]i16 = undefined,
    depth: usize = 0,

    pub fn init(data: []const u8) Reader {
        return .{ .data = data };
    }

    fn byte(self: *Reader) !u8 {
        if (self.at >= self.data.len) return Error.TruncatedMetadata;
        const b = self.data[self.at];
        self.at += 1;
        return b;
    }

    fn varint(self: *Reader) !u64 {
        var out: u64 = 0;
        var shift: u6 = 0;
        while (true) {
            const b = try self.byte();
            out |= @as(u64, b & 0x7f) << shift;
            if (b & 0x80 == 0) return out;
            if (shift >= 63 - 7) return Error.TruncatedMetadata;
            shift += 7;
        }
    }

    fn zigzag(self: *Reader) !i64 {
        const n = try self.varint();
        return @as(i64, @bitCast(n >> 1)) ^ -@as(i64, @intCast(n & 1));
    }

    /// Enters a struct. Field ids are relative to the enclosing struct's last id,
    /// so every `structBegin` needs its `structEnd`.
    pub fn structBegin(self: *Reader) void {
        if (self.depth < self.stack.len) self.stack[self.depth] = self.last_id;
        self.depth += 1;
        self.last_id = 0;
    }

    pub fn structEnd(self: *Reader) void {
        if (self.depth > 0) self.depth -= 1;
        self.last_id = if (self.depth < self.stack.len) self.stack[self.depth] else 0;
    }

    /// The next field of the current struct, or null at its stop byte.
    pub fn field(self: *Reader) !?FieldHeader {
        const header = try self.byte();
        if (header == STOP) return null;
        const kind = header & 0x0f;
        const delta: i16 = @intCast(header >> 4);
        const id = if (delta == 0) @as(i16, @intCast(try self.zigzag())) else self.last_id + delta;
        self.last_id = id;
        return .{ .id = id, .kind = kind };
    }

    pub fn i32v(self: *Reader) !i32 {
        return @truncate(try self.zigzag());
    }

    pub fn i64v(self: *Reader) !i64 {
        return self.zigzag();
    }

    pub fn binary(self: *Reader) ![]const u8 {
        const len: usize = @intCast(try self.varint());
        if (self.at + len > self.data.len) return Error.TruncatedMetadata;
        const out = self.data[self.at..][0..len];
        self.at += len;
        return out;
    }

    /// A list header: how many elements, and of what.
    pub fn listBegin(self: *Reader) !struct { usize, u8 } {
        const header = try self.byte();
        const kind = header & 0x0f;
        var size: usize = header >> 4;
        if (size == 15) size = @intCast(try self.varint());
        return .{ size, kind };
    }

    /// Steps over a value of `kind` without interpreting it. This is what lets a
    /// file written by a newer library, carrying fields this reader has never
    /// heard of, still be read.
    pub fn skip(self: *Reader, kind: u8) !void {
        switch (kind) {
            TRUE, FALSE => {},
            BYTE => _ = try self.byte(),
            I16, I32, I64 => _ = try self.zigzag(),
            DOUBLE => {
                if (self.at + 8 > self.data.len) return Error.TruncatedMetadata;
                self.at += 8;
            },
            BINARY => _ = try self.binary(),
            LIST, SET => {
                const n, const element = try self.listBegin();
                for (0..n) |_| try self.skip(element);
            },
            MAP => {
                const n: usize = @intCast(try self.varint());
                if (n == 0) return;
                const kinds = try self.byte();
                for (0..n) |_| {
                    try self.skip(kinds >> 4);
                    try self.skip(kinds & 0x0f);
                }
            },
            STRUCT => {
                self.structBegin();
                while (try self.field()) |fh| try self.skip(fh.kind);
                self.structEnd();
            },
            else => return Error.UnknownThriftType,
        }
    }

    /// How far the cursor has got, which a page header reader needs to know where
    /// the page's own bytes begin.
    pub fn position(self: Reader) usize {
        return self.at;
    }
};

test "fields are read by delta and values unzigzagged" {
    // {1: 7, 2: "hi", 3: [4, 5]} in compact form, then a stop byte.
    const message = [_]u8{ 0x15, 0x0e, 0x18, 0x02, 'h', 'i', 0x19, 0x25, 0x08, 0x0a, 0x00 };
    var r = Reader.init(&message);
    r.structBegin();
    var seen: usize = 0;
    while (try r.field()) |fh| {
        switch (fh.id) {
            1 => try std.testing.expectEqual(@as(i32, 7), try r.i32v()),
            2 => try std.testing.expectEqualStrings("hi", try r.binary()),
            3 => {
                const n, _ = try r.listBegin();
                try std.testing.expectEqual(@as(usize, 2), n);
                try std.testing.expectEqual(@as(i32, 4), try r.i32v());
                try std.testing.expectEqual(@as(i32, 5), try r.i32v());
            },
            else => try r.skip(fh.kind),
        }
        seen += 1;
    }
    r.structEnd();
    try std.testing.expectEqual(@as(usize, 3), seen);
}

test "a field nobody wants is stepped over" {
    const message = [_]u8{ 0x15, 0x0e, 0x18, 0x02, 'h', 'i', 0x19, 0x25, 0x08, 0x0a, 0x00 };
    var r = Reader.init(&message);
    r.structBegin();
    var last: ?i32 = null;
    while (try r.field()) |fh| {
        if (fh.id == 3) {
            _ = try r.listBegin();
            _ = try r.i32v();
            last = try r.i32v();
        } else try r.skip(fh.kind);
    }
    r.structEnd();
    try std.testing.expectEqual(@as(?i32, 5), last);
}
